src/annoylib.h

```diff
@@ -1172,7 +1199,87 @@ template<typename S, typename T, typename Distance, typename Random>
     return get_node_ptr<S, Node>(_nodes, _s, i);
   }
 
-  S _make_tree(const vector<S >& indices, bool is_root) {
+  void _thread_build(int q, int thread_idx, ThreadBarrier& barrier, std::mutex& nodes_mutex) {
+    Random _random;
+    // Each thread needs its own seed, otherwise each thread would be building the same tree(s)
+    int seed = _is_seeded ? _seed + thread_idx : thread_idx;
+    _random.set_seed(seed);
+
+    vector<vector<Node*> > thread_trees;
+    vector<S> thread_roots;
+    while (1) {
+      if (q == -1) {
+        size_t thread_n_nodes = 0;
+        for (size_t tree_idx = 0; tree_idx < thread_trees.size(); tree_idx++) {
+          thread_n_nodes += thread_trees[tree_idx].size();
+        }
+        if (thread_n_nodes >= 2 * (size_t)_n_items) {
+          break;
+        }
+      } else {
+        if (thread_roots.size() >= (size_t)q) {
+          break;
+        }
+      }
+
+      if (_verbose) showUpdate("pass %zd...\n", thread_roots.size());
+
+      vector<S> indices;
+      for (S i = 0; i < _n_items; i++) {
+        if (_get(i)->n_descendants >= 1) // Issue #223
+          indices.push_back(i);
+      }
+
+      vector<Node*> split_nodes;
+      // Each thread is essentially pretending to build only one tree that will get inserted
+      // right after the already inserted items. Indices of split nodes start with _n_items, n_items + 1, ...
+      // We do not want to mutate the _nodes array during tree construction due to reallocation issues. That is
+      // why each thread stores the trees locally until all threads are ready to insert them into _nodes.
+      S root_node = _make_tree(indices, split_nodes, true, _random);
+      thread_roots.push_back(root_node);
+      thread_trees.push_back(split_nodes);
+    }
+
+    // Wait for all threads to finish before we can start inserting tree nodes into global _nodes array
+    barrier.wait();
```

Comment:
isn't locking the mutex enough to enforce this?

Reply:
Unfortunately not, due to reasons noted here: https://github.com/spotify/annoy/pull/495#discussion_r461732904 We have to wait for all threads finish their work before we can mutate the global _nodes buffer.